File: clients/python/src/varta/_transport.py
"""Transport abstraction for the Varta agent.

The :class:`BeatTransport` ABC mirrors the Rust trait of the same name
(`crates/varta-client/src/transport.rs`). Concrete implementations:

* :class:`UdsTransport` — Unix domain datagram socket.
* :class:`UdpTransport` — connected UDP socket.
* :class:`SecureUdpTransport` — UDP + ChaCha20-Poly1305 AEAD (requires
  ``cryptography``; raises :class:`CryptographyMissingError` on use).

Every transport must guarantee:

1. Non-blocking I/O at construction (``setblocking(False)``).
2. ``send(buf)`` accepts a 32-byte ``bytes``-like, returns ``FRAME_BYTES``
   only after a full logical heartbeat is accepted, and raises ``OSError``
   on kernel-level failure.
3. ``reconnect()`` rebuilds the underlying socket from the original
   construction parameters. Cold path; allocation is allowed.
"""

from __future__ import annotations

import abc
import errno
import os
import socket
from typing import Optional, Tuple, Union

from ._vlp import FRAME_BYTES
from ._vlp_secure import (
    KEY_BYTES,
    SECURE_MASTER_BYTES,
    SECURE_SHARED_BYTES,
    derive_iv_prefix,
    encode_master,
    encode_shared,
)

__all__ = [
    "BeatTransport",
    "UdsTransport",
    "UdpTransport",
    "SecureUdpTransport",
]


def _closed_socket_error() -> OSError:
    return OSError(errno.EBADF, "transport is closed")


class BeatTransport(abc.ABC):
    """Transport contract for :class:`varta.Varta`.

    Implementations must be non-blocking; the agent layer translates
    :class:`BlockingIOError` and other ``OSError`` subclasses into the
    :class:`varta.BeatOutcome` taxonomy via
    :func:`varta.classify_send_error`.
    """

    @abc.abstractmethod
    def send(self, buf: bytes) -> int:
        """Send the 32-byte frame ``buf`` over the underlying socket.

        Returns ``FRAME_BYTES`` on full logical success. Positive short writes
        are not successful heartbeats; the agent treats any other return value
        as ``Failed(WriteZero)``. Raises ``OSError`` (including
        ``BlockingIOError``) on failure.
        """

    @abc.abstractmethod
    def reconnect(self) -> None:
        """Rebuild the socket without destroying the current one on failure."""

    def close(self) -> None:  # pragma: no cover - default is no-op
        """Best-effort cleanup; subclasses override if they own a socket."""


# ---------------------------------------------------------------------------
# UDS
# ---------------------------------------------------------------------------


def _uds_socket(path: str) -> socket.socket:
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
    try:
        sock.setblocking(False)
        sock.connect(path)
    except OSError:
        sock.close()
        raise
    return sock


class UdsTransport(BeatTransport):
    """Unix-domain-datagram transport. Default for ``Varta.connect``."""

    __slots__ = ("_path", "_sock")

    def __init__(self, path: Union[str, "os.PathLike[str]"]) -> None:
        self._path = os.fspath(path)
        self._sock: Optional[socket.socket] = _uds_socket(self._path)

    def send(self, buf: bytes) -> int:
        sock = self._sock
        if sock is None:
            raise _closed_socket_error()
        return sock.send(buf)

    def reconnect(self) -> None:
        sock = _uds_socket(self._path)
        old = self._sock
        self._sock = sock
        if old is not None:
            try:
                old.close()
            except OSError:
                pass

    def close(self) -> None:
        if self._sock is not None:
            self._sock.close()
            self._sock = None

    def __del__(self) -> None:  # pragma: no cover - GC path
        try:
            self.close()
        except Exception:
            pass


# ---------------------------------------------------------------------------
# UDP
# ---------------------------------------------------------------------------


Address = Tuple[str, int]


def _udp_socket(addr: Address) -> socket.socket:
    family = socket.AF_INET6 if ":" in addr[0] else socket.AF_INET
    sock = socket.socket(family, socket.SOCK_DGRAM)
    try:
        sock.setblocking(False)
        sock.bind(("", 0) if family == socket.AF_INET else ("", 0, 0, 0))
        sock.connect(addr)
    except OSError:
        sock.close()
        raise
    return sock


class UdpTransport(BeatTransport):
    """Connected UDP datagram transport (cleartext)."""

    __slots__ = ("_addr", "_sock")

    def __init__(self, addr: Address) -> None:
        self._addr = addr
        self._sock: Optional[socket.socket] = _udp_socket(self._addr)

    def send(self, buf: bytes) -> int:
        sock = self._sock
        if sock is None:
            raise _closed_socket_error()
        return sock.send(buf)

    def reconnect(self) -> None:
        sock = _udp_socket(self._addr)
        old = self._sock
        self._sock = sock
        if old is not None:
            try:
                old.close()
            except OSError:
                pass

    def close(self) -> None:
        if self._sock is not None:
            self._sock.close()
            self._sock = None

    def __del__(self) -> None:  # pragma: no cover
        try:
            self.close()
        except Exception:
            pass


# ---------------------------------------------------------------------------
# Secure UDP — ChaCha20-Poly1305 AEAD.
# ---------------------------------------------------------------------------

# A 32-bit AEAD counter exhausts at 2^32. The Rust transport rotates the
# IV prefix when the counter is about to wrap; we mirror that behaviour
# so cross-implementation traces stay aligned.
_AEAD_COUNTER_LIMIT = 0xFFFFFFFF


class SecureUdpTransport(BeatTransport):
    """ChaCha20-Poly1305 AEAD over UDP.

    Two construction modes:

    * **Shared key** — both observer and agent are configured with the
      same 32-byte pre-shared key. Wire frame is 60 bytes.
    * **Master key** — observer holds a master key; agent passes its
      PID to derive a per-agent key via HKDF. Wire frame is 64 bytes
      (4-byte PID prefix is AAD).

    Session IV state (16-byte salt + 8-byte derived prefix + 32-bit
    counter) lives in process memory. Salt is read from :func:`os.urandom`
    at connect time and after any :meth:`reconnect` — the structural
    guarantee against AEAD nonce reuse across ``fork(2)``.
    """

    __slots__ = (
        "_addr",
        "_key",
        "_master_key",
        "_sock",
        "_session_salt",
        "_iv_prefix",
        "_iv_prefix_index",
        "_iv_counter",
    )

    def __init__(
        self,
        addr: Address,
        *,
        key: Optional[bytes] = None,
        master_key: Optional[bytes] = None,
    ) -> None:
        if (key is None) == (master_key is None):
            raise ValueError("provide exactly one of key or master_key")
        if key is not None and len(key) != KEY_BYTES:
            raise ValueError(f"key must be {KEY_BYTES} bytes")
        if master_key is not None and len(master_key) != KEY_BYTES:
            raise ValueError(f"master_key must be {KEY_BYTES} bytes")
        self._addr = addr
        self._key = key
        self._master_key = master_key
        self._sock: Optional[socket.socket] = None
        self._session_salt = b""
        self._iv_prefix = b""
        self._iv_prefix_index = 0
        self._iv_counter = 0
        self._open()

    def _open(self) -> None:
        sock, session_salt, iv_prefix = self._prepare_session()
        self._sock = sock
        self._session_salt = session_salt
        self._iv_prefix = iv_prefix
        self._iv_prefix_index = 0
        self._iv_counter = 0

    def _prepare_session(self) -> Tuple[socket.socket, bytes, bytes]:
        sock = _udp_socket(self._addr)
        try:
            session_salt = os.urandom(16)
            iv_prefix = derive_iv_prefix(session_salt, 0)
        except BaseException:
            try:
                sock.close()
            except OSError:
                pass
            raise
        return sock, session_salt, iv_prefix

    def send(self, buf: bytes) -> int:
        sock = self._sock
        if sock is None:
            raise _closed_socket_error()
        if len(buf) != FRAME_BYTES:
            raise ValueError("secure-UDP transport expects a 32-byte plaintext frame")
        # Compute the nonce-wrap into locals; transport state is mutated only
        # after the kernel accepts the datagram (commit-on-success), so a
        # Dropped send never advances the prefix index or resets the counter.
        # The prior code rotated the prefix BEFORE the syscall, leaving the
        # prefix rotated when a non-blocking send raised BlockingIOError at the
        # wrap boundary (the Python sibling of the Go fix).
        prefix_index = self._iv_prefix_index
        counter = self._iv_counter
        iv_prefix = self._iv_prefix
        if counter >= _AEAD_COUNTER_LIMIT:
            if prefix_index >= _AEAD_COUNTER_LIMIT:
                self.reconnect()
                sock = self._sock
                if sock is None:
                    raise _closed_socket_error()
                prefix_index = self._iv_prefix_index
                counter = self._iv_counter
                iv_prefix = self._iv_prefix
            else:
                prefix_index += 1
                counter = 0
                iv_prefix = derive_iv_prefix(self._session_salt, prefix_index)
        if self._master_key is not None:
            agent_pid = os.getpid() & 0xFFFFFFFF
            wire = encode_master(
                self._master_key, agent_pid, iv_prefix, counter, buf
            )
            assert len(wire) == SECURE_MASTER_BYTES
        else:
            assert self._key is not None
            wire = encode_shared(self._key, iv_prefix, counter, buf)
            assert len(wire) == SECURE_SHARED_BYTES
        sent = sock.send(wire)
        if sent != len(wire):
            return 0
        # Commit-on-success: only advance state after the kernel accepts the
        # full encrypted datagram. A Dropped or short send must not consume a
        # nonce or burn a prefix.
        self._iv_prefix_index = prefix_index
        self._iv_prefix = iv_prefix
        self._iv_counter = counter + 1
        return FRAME_BYTES

    def reconnect(self) -> None:
        sock, session_salt, iv_prefix = self._prepare_session()
        old = self._sock
        self._sock = sock
        self._session_salt = session_salt
        self._iv_prefix = iv_prefix
        self._iv_prefix_index = 0
        self._iv_counter = 0
        if old is not None:
            try:
                old.close()
            except OSError:
                pass

    def close(self) -> None:
        if self._sock is not None:
            self._sock.close()
            self._sock = None

    # Test hooks (parity with Rust `set_iv_counter_for_test` etc.).
    def _set_iv_counter_for_test(self, value: int) -> None:
        self._iv_counter = value & 0xFFFFFFFF

    def _set_iv_prefix_index_for_test(self, value: int) -> None:
        self._iv_prefix_index = value & 0xFFFFFFFF
        self._iv_prefix = derive_iv_prefix(self._session_salt, self._iv_prefix_index)

    def _iv_prefix_for_test(self) -> bytes:
        return self._iv_prefix

    def _iv_prefix_index_for_test(self) -> int:
        return self._iv_prefix_index

    def _iv_counter_for_test(self) -> int:
        return self._iv_counter

    def __del__(self) -> None:  # pragma: no cover
        try:
            self.close()
        except Exception:
            pass

File: clients/node/src/transport.ts
// Beat transport abstractions — UDS (AF_UNIX/SOCK_DGRAM), plaintext
// UDP, and ChaCha20-Poly1305-AEAD UDP.
//
// The UDS transport relies on the `node-unix-socket` napi-rs addon
// (listed under `optionalDependencies`). It ships prebuilds for
// darwin-x64/arm64 and linux-x64/arm64 (gnu + musl); installing on a
// platform without a published prebuild succeeds with the addon
// absent, and the UDS transport raises `UdsUnavailableError` at
// construction. UDP / secure-UDP work on every Node platform with
// only the stdlib.
//
// UDP and secure-UDP use connected-mode sockets (`dgram.Socket.connect`)
// so the kernel's ICMP `port unreachable` is queued onto the socket's
// `error` event and surfaces as `DropReason.NoObserver` on a subsequent
// beat. Pre-`connect` event sends are queued internally by libuv;
// errors from any source (send callback, recvmsg ICMP, async connect)
// drain into `pendingError` for the next caller.

import { randomBytes } from "node:crypto";
import { createSocket, type Socket } from "node:dgram";
import { createRequire } from "node:module";

import {
  encodeMaster,
  encodeShared,
  IV_RANDOM_BYTES,
  PLAINTEXT_BYTES,
  SESSION_SALT_BYTES,
  deriveIvPrefix,
} from "./vlp_secure.js";

export interface BeatTransport {
  // Synchronous send. Returns nothing on success; throws a
  // `NodeJS.ErrnoException`-shaped error on failure. libuv may report
  // kernel-level errors on a later send's callback or via a recvmsg
  // ICMP error event; transports drain those into `pendingError` so
  // callers see Rust-style synchronous semantics.
  send(buf: Buffer): void;
  // Rebuild the transport with the strong exception guarantee: if
  // replacement setup throws, the current socket and any secure-session
  // state remain usable and unchanged.
  reconnect(): void;
  close(): void;
}

function takePendingError(holder: { pendingError: NodeJS.ErrnoException | null }):
  | NodeJS.ErrnoException
  | null {
  const e = holder.pendingError;
  holder.pendingError = null;
  return e;
}

// ─── UDS (AF_UNIX/SOCK_DGRAM) ──────────────────────────────────

export class UdsUnavailableError extends Error {
  readonly cause: unknown;
  constructor(cause: unknown) {
    super(
      "UDS transport requires the optional `node-unix-socket` addon. " +
        "Install it with `npm install node-unix-socket`, or fall back to " +
        "`Varta.connectUdp(\"127.0.0.1\", port)` on platforms without a prebuild.",
    );
    this.name = "UdsUnavailableError";
    this.cause = cause;
  }
}

// `node-unix-socket`'s `DgramSocket` reports errors without populating
// `err.code` / `err.errno`. Translate the human-readable message back
// into the symbolic form that `classifySendError` expects.
function normalizeUdsError(err: unknown): NodeJS.ErrnoException {
  const e = (err instanceof Error ? err : new Error(String(err))) as NodeJS.ErrnoException;
  if (typeof e.code === "string" && e.code.length > 0) return e;
  const msg = (e.message ?? "").toLowerCase();
  if (msg.includes("no such file")) e.code = "ENOENT";
  else if (msg.includes("connection refused")) e.code = "ECONNREFUSED";
  else if (msg.includes("no buffer space")) e.code = "ENOBUFS";
  else if (msg.includes("resource temporarily unavailable")) e.code = "EAGAIN";
  else if (msg.includes("would block")) e.code = "EWOULDBLOCK";
  else if (msg.includes("no space left")) e.code = "ENOSPC";
  else if (msg.includes("broken pipe")) e.code = "EPIPE";
  else if (msg.includes("connection reset")) e.code = "ECONNRESET";
  else if (msg.includes("transport endpoint is not connected")) e.code = "ENOTCONN";
  return e;
}

interface DgramSocketLike {
  bind(socketPath: string): void;
  sendTo(
    buf: Buffer,
    offset: number,
    length: number,
    destPath: string,
    onWrite?: (err: undefined | Error) => void,
  ): void;
  close(): void;
  on(event: "error", listener: (err: Error) => void): unknown;
  on(event: "data", listener: (buf: Buffer, path: string) => void): unknown;
}

interface NodeUnixSocketModule {
  DgramSocket: new () => DgramSocketLike;
}

let cachedNodeUnixSocket: NodeUnixSocketModule | null | "missing" = null;

const requireFromHere = createRequire(import.meta.url);

function loadNodeUnixSocket(): NodeUnixSocketModule {
  if (cachedNodeUnixSocket === "missing") {
    throw new UdsUnavailableError(new Error("module not installed"));
  }
  if (cachedNodeUnixSocket !== null) return cachedNodeUnixSocket;
  try {
    // `createRequire` keeps this importable from an ESM build without a
    // top-level `await import` (which would force every consumer to
    // tolerate an async load even when they never touch UDS).
    const mod = requireFromHere("node-unix-socket") as NodeUnixSocketModule;
    cachedNodeUnixSocket = mod;
    return mod;
  } catch (err) {
    cachedNodeUnixSocket = "missing";
    throw new UdsUnavailableError(err);
  }
}

export class UdsTransport implements BeatTransport {
  private socket: DgramSocketLike;
  private readonly path: string;
  pendingError: NodeJS.ErrnoException | null = null;

  constructor(path: string) {
    this.path = path;
    this.socket = this.openSocket();
  }

  private openSocket(): DgramSocketLike {
    const { DgramSocket } = loadNodeUnixSocket();
    const s = new DgramSocket();
    // Upstream's `onError` self-closes the socket and re-emits as
    // `error`. Drain into pendingError so the next beat surfaces it,
    // and rely on `reconnect()` to rebuild. Most sendTo failures
    // arrive on the per-call callback (not this event), so this path
    // is rare.
    s.on("error", (err) => {
      if (this.socket === s) {
        this.pendingError = normalizeUdsError(err);
      }
    });
    return s;
  }

  send(buf: Buffer): void {
    const queued = takePendingError(this);
    if (queued !== null) throw queued;
    // Copy the caller's scratch buffer before handing off to libuv —
    // matching the UDP transports' guard. The agent reuses a single
    // 32-byte buffer across beats.
    const owned = Buffer.from(buf);
    const socket = this.socket;
    socket.sendTo(owned, 0, owned.length, this.path, (err) => {
      if (this.socket === socket && err !== null && err !== undefined) {
        this.pendingError = normalizeUdsError(err);
      }
    });
  }

  reconnect(): void {
    const replacement = this.openSocket();
    const old = this.socket;
    this.socket = replacement;
    this.pendingError = null;
    try {
      old.close();
    } catch {
      // Already closed — fine.
    }
  }

  close(): void {
    try {
      this.socket.close();
    } catch {
      // Already closed.
    }
  }
}

// ─── Plaintext UDP (connected-mode) ────────────────────────────

// Bound on how many beats may queue while `socket.connect` is in
// flight. IPv4 numeric connect completes in one libuv tick, so the
// queue is normally empty by the second beat; this cap exists only to
// guard against a `connect` that never fires (peer DNS issues etc.).
const PRE_CONNECT_QUEUE_LIMIT = 64;
const U32_MAX = 0xffffffff;

export class UdpTransport implements BeatTransport {
  private socket: Socket;
  private readonly host: string;
  private readonly port: number;
  private connected = false;
  private preConnectQueue: Buffer[] = [];
  pendingError: NodeJS.ErrnoException | null = null;

  constructor(host: string, port: number) {
    this.host = host;
    this.port = port;
    this.socket = this.openSocket();
  }

  private openSocket(): Socket {
    const family: "udp4" | "udp6" = this.host.includes(":") ? "udp6" : "udp4";
    const s = createSocket(family);
    // ICMP `port unreachable` for a connected dgram socket lands here
    // (libuv's recvmsg path), as do async-connect failures and any
    // out-of-band socket errors. Swallow into pendingError so the next
    // `send()` call surfaces it instead of crashing the process.
    s.on("error", (err) => {
      if (this.socket === s) {
        this.pendingError = err as NodeJS.ErrnoException;
      }
    });
    s.unref();
    // Connect for ICMP error propagation. Node's `dgram.Socket.send`
    // rejects the connected-mode (no-port/host) signature while the
    // socket is still CONNECTING; queue pre-connect-event beats and
    // flush them once the `connect` callback fires.
    try {
      s.connect(this.port, this.host, () => {
        if (this.socket !== s) return;
        this.connected = true;
        while (this.preConnectQueue.length > 0) {
          const owned = this.preConnectQueue.shift()!;
          s.send(owned, (err) => {
            if (this.socket === s && err !== null && err !== undefined) {
              this.pendingError = err as NodeJS.ErrnoException;
            }
          });
        }
      });
    } catch (err) {
      try {
        s.close();
      } catch {
        // Socket never became active.
      }
      throw err;
    }
    return s;
  }

  send(buf: Buffer): void {
    const queued = takePendingError(this);
    if (queued !== null) throw queued;
    // Buffer copy: libuv does NOT copy before handing to the kernel,
    // and the caller reuses a single 32-byte scratch buffer.
    const owned = Buffer.from(buf);
    if (!this.connected) {
      if (this.preConnectQueue.length >= PRE_CONNECT_QUEUE_LIMIT) {
        const e: NodeJS.ErrnoException = new Error(
          "UdpTransport: pre-connect queue full",
        );
        e.code = "ENOBUFS";
        throw e;
      }
      this.preConnectQueue.push(owned);
      return;
    }
    const socket = this.socket;
    socket.send(owned, (err) => {
      if (this.socket === socket && err !== null && err !== undefined) {
        this.pendingError = err as NodeJS.ErrnoException;
      }
    });
  }

  reconnect(): void {
    const replacement = this.openSocket();
    const old = this.socket;
    this.socket = replacement;
    this.connected = false;
    this.pendingError = null;
    // Keep any beats that were queued while the previous socket was still
    // connecting. They have not been sent yet, so discarding them would burn
    // committed client state without giving the replacement socket a chance
    // to deliver them.
    try {
      old.close();
    } catch {
      // Already closed — fine.
    }
  }

  close(): void {
    try {
      this.socket.close();
    } catch {
      // Already closed.
    }
  }
}

// ─── Secure UDP (ChaCha20-Poly1305 AEAD, connected-mode) ───────

export type SecureUdpKind = "shared" | "master";

interface SecureKeyShared {
  kind: "shared";
  key: Buffer;
}
interface SecureKeyMaster {
  kind: "master";
  masterKey: Buffer;
}
type SecureKey = SecureKeyShared | SecureKeyMaster;

interface NonceReservation {
  ivRandom: Buffer;
  counter: number;
  nextPrefixIndex: number;
  nextIvPrefix: Buffer;
  nextCounter: number;
}

export class SecureUdpTransport implements BeatTransport {
  private socket: Socket;
  private readonly host: string;
  private readonly port: number;
  private readonly secret: SecureKey;
  private sessionSalt: Buffer;
  private ivPrefix: Buffer;
  private prefixIndex: number;
  private counter: number;
  private connected = false;
  private preConnectQueue: Buffer[] = [];
  pendingError: NodeJS.ErrnoException | null = null;

  private constructor(host: string, port: number, secret: SecureKey) {
    this.host = host;
    this.port = port;
    this.secret = secret;
    this.sessionSalt = randomBytes(SESSION_SALT_BYTES);
    this.prefixIndex = 0;
    this.counter = 0;
    this.ivPrefix = deriveIvPrefix(this.sessionSalt, this.prefixIndex);
    this.socket = this.openSocket();
  }

  static shared(host: string, port: number, key: Buffer): SecureUdpTransport {
    if (key.length !== 32) {
      throw new RangeError("secure-UDP shared key must be 32 bytes");
    }
    return new SecureUdpTransport(host, port, { kind: "shared", key: Buffer.from(key) });
  }

  static master(host: string, port: number, masterKey: Buffer): SecureUdpTransport {
    if (masterKey.length !== 32) {
      throw new RangeError("secure-UDP master key must be 32 bytes");
    }
    return new SecureUdpTransport(host, port, {
      kind: "master",
      masterKey: Buffer.from(masterKey),
    });
  }

  private openSocket(): Socket {
    const family: "udp4" | "udp6" = this.host.includes(":") ? "udp6" : "udp4";
    const s = createSocket(family);
    s.on("error", (err) => {
      if (this.socket === s) {
        this.pendingError = err as NodeJS.ErrnoException;
      }
    });
    s.unref();
    try {
      s.connect(this.port, this.host, () => {
        if (this.socket !== s) return;
        this.connected = true;
        while (this.preConnectQueue.length > 0) {
          const wire = this.preConnectQueue.shift()!;
          s.send(wire, (err) => {
            if (this.socket === s && err !== null && err !== undefined) {
              this.pendingError = err as NodeJS.ErrnoException;
            }
          });
        }
      });
    } catch (err) {
      try {
        s.close();
      } catch {
        // Socket never became active.
      }
      throw err;
    }
    return s;
  }

  // Test hooks (matching Python `_set_*_for_test` and Go `*ForTest`).
  __setCounterForTest(v: number): void {
    this.counter = v >>> 0;
  }
  __setPrefixIndexForTest(v: number): void {
    this.prefixIndex = v >>> 0;
    this.ivPrefix = deriveIvPrefix(this.sessionSalt, this.prefixIndex);
  }
  __getCounterForTest(): number {
    return this.counter;
  }
  __getPrefixIndexForTest(): number {
    return this.prefixIndex;
  }
  __getIvPrefixForTest(): Buffer {
    return Buffer.from(this.ivPrefix);
  }

  private prepareNonce(): NonceReservation {
    if (this.counter === U32_MAX && this.prefixIndex === U32_MAX) {
      // The 64-bit per-session nonce space is exhausted. Match the Rust
      // reference: rotate the whole secure session before emitting another
      // frame, rather than wrapping back to prefix index 0 under the same salt.
      this.reconnect();
    }

    let nextPrefixIndex = this.prefixIndex;
    let nextIvPrefix = this.ivPrefix;
    let counter = this.counter;
    if (counter === U32_MAX) {
      nextPrefixIndex = (nextPrefixIndex + 1) >>> 0;
      nextIvPrefix = deriveIvPrefix(this.sessionSalt, nextPrefixIndex);
      counter = 0;
    }

    return {
      ivRandom: Buffer.from(nextIvPrefix.subarray(0, IV_RANDOM_BYTES)),
      counter,
      nextPrefixIndex,
      nextIvPrefix,
      nextCounter: (counter + 1) >>> 0,
    };
  }

  private commitNonce(reservation: NonceReservation): void {
    this.prefixIndex = reservation.nextPrefixIndex;
    this.ivPrefix = reservation.nextIvPrefix;
    this.counter = reservation.nextCounter;
  }

  send(buf: Buffer): void {
    const queued = takePendingError(this);
    if (queued !== null) throw queued;
    if (buf.length !== PLAINTEXT_BYTES) {
      throw new RangeError("plaintext must be a 32-byte VLP frame");
    }
    if (!this.connected && this.preConnectQueue.length >= PRE_CONNECT_QUEUE_LIMIT) {
      const e: NodeJS.ErrnoException = new Error(
        "SecureUdpTransport: pre-connect queue full",
      );
      e.code = "ENOBUFS";
      throw e;
    }

    // Compute IV state into locals and commit only after the frame is accepted
    // for delivery (or queued during connect). A later libuv callback error
    // still burns the nonce slot because the datagram may already have escaped.
    const nonce = this.prepareNonce();

    let wire: Buffer;
    if (this.secret.kind === "shared") {
      wire = encodeShared(this.secret.key, nonce.ivRandom, nonce.counter, buf);
    } else {
      wire = encodeMaster(
        this.secret.masterKey,
        process.pid >>> 0,
        nonce.ivRandom,
        nonce.counter,
        buf,
      );
    }

    if (!this.connected) {
      this.preConnectQueue.push(wire);
      this.commitNonce(nonce);
      return;
    }
    const socket = this.socket;
    socket.send(wire, (err) => {
      if (this.socket === socket && err !== null && err !== undefined) {
        this.pendingError = err as NodeJS.ErrnoException;
      }
    });
    this.commitNonce(nonce);
  }

  reconnect(): void {
    const newSalt = randomBytes(SESSION_SALT_BYTES);
    const newPrefixIndex = 0;
    const newIvPrefix = deriveIvPrefix(newSalt, newPrefixIndex);
    const newSocket = this.openSocket();
    const old = this.socket;
    this.socket = newSocket;
    this.connected = false;
    this.pendingError = null;
    // Preserve queued beats across reconnect for the same reason as the UDP
    // transport: a queued pre-connect datagram has not been accepted yet.
    this.sessionSalt = newSalt;
    this.prefixIndex = newPrefixIndex;
    this.counter = 0;
    this.ivPrefix = newIvPrefix;
    try {
      old.close();
    } catch {
      // Already closed.
    }
  }

  close(): void {
    try {
      this.socket.close();
    } catch {
      // Already closed.
    }
  }
}
